portfolio/src/main/java/com/google/sps/servlets/ImageFormHandlerServlet.java

```diff
@@ -0,0 +1,103 @@
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import com.google.gson.Gson;
+
+/**
+ * When the user submits the form, Blobstore processes the file upload and then forwards the request
+ * to this servlet. This servlet can then process the request using the file URL we get from
+ * Blobstore, but in this case the image URL is put into an array to be used to appear on the webpage.
+ */
+@WebServlet("/my-form-handler")
+public class ImageFormHandlerServlet extends HttpServlet {
+  private List<String> images = new ArrayList<>();
+  // Variable to be used to allow multiple images to be uploaded,
+  // and allow the image on the page to refresh, but only one image
+  //is ever stored in the array 
+  private int maxNumberOfImages = 2;
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Gson gson = new Gson();
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(images));
+  }
+  
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    String imageUrl = getUploadedFileUrl(request, "image");
+    if(imageUrl != null){
+      images.add(imageUrl);
+    }
+
+    if(images.size() >= maxNumberOfImages){
+      images.remove(0);
+    }
+    
+    response.sendRedirect("/index.html");
+  }
+
+  // Returns a URL that points to the uploaded file, or null if the user didn't upload a file.
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get("image");
+    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    BlobKey blobKey = blobKeys.get(0);
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+    
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
+    ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+
+    try {
+      URL url = new URL(imagesService.getServingUrl(options));
+      return url.getPath();
+    } catch (MalformedURLException e) {
+      return imagesService.getServingUrl(options);
+    }
```

Comment:
Can you add a comment explaining what the goal is here? It's not clear why the servingUrl is a good fallback here. And what are the implications for the URL constructor failing?

Reply:
This is from the example they gave us, the reason for the try catch as to run this locally, the path to the image has to be relative for the image to load, as imagesService contains a host for the image to load.

Reply:
Can you check this again